p2p/src/sync/mod.rs

```diff
@@ -89,445 +97,485 @@ where
-            Err(err) => Err(P2pError::ChainstateError(err)),
+            return Err(P2pError::ProtocolError(ProtocolError::DisconnectedHeaders));
         }
-    }
 
-    /// Process incoming header response
-    pub async fn process_header_response(
-        &mut self,
-        peer_id: T::PeerId,
-        headers: Vec<BlockHeader>,
-    ) -> crate::Result<()> {
-        match self.validate_header_response(&peer_id, headers).await {
-            Ok(Some(header)) => self.send_block_request(peer_id, header.get_id()),
-            Ok(None) => {
-                self.peers
-                    .get_mut(&peer_id)
-                    .ok_or(P2pError::PeerError(PeerError::PeerDoesntExist))?
-                    .set_state(peer::PeerSyncState::Idle);
-                Ok(())
+        let is_max_headers =
+            headers.len() == Into::<usize>::into(self.p2p_config.msg_header_count_limit.clone());
+        let headers = self
+            .chainstate_handle
+            .call(|c| c.filter_already_existing_blocks(headers))
+            .await??;
+        if headers.is_empty() {
+            // A peer can have more headers if we have received the maximum amount of them.
+            if is_max_headers {
+                self.request_headers(peer).await?;
             }
-            Err(err) => Err(err),
+            return Ok(());
         }
+
+        // Only the first header can be checked with the `preliminary_header_check` function.
+        let first_header = headers
+            .first()
+            // This is OK because of the `headers.is_empty()` check above.
+            .expect("Headers shouldn't be empty")
+            .clone();
+        self.chainstate_handle
+            .call(|c| c.preliminary_header_check(first_header))
+            .await??;
+
+        self.request_blocks(peer, headers)
     }
 
-    /// Validate incoming block response
-    async fn validate_block_response(
-        &mut self,
-        peer_id: &T::PeerId,
-        blocks: Vec<Block>,
-    ) -> crate::Result<Option<BlockHeader>> {
-        let peer = self
+    // TODO: This shouldn't be public.
+    pub async fn handle_block_response(&mut self, peer: T::PeerId, block: Block) -> Result<()> {
+        log::debug!("Block ({}) from peer {peer}", block.get_id());
+
+        let peer_state = self
             .peers
-            .get_mut(peer_id)
+            .get_mut(&peer)
             .ok_or(P2pError::PeerError(PeerError::PeerDoesntExist))?;
+        if peer_state.requested_blocks.take(&block.get_id()).is_none() {
+            return Err(P2pError::ProtocolError(ProtocolError::UnexpectedMessage(
+                "block response",
+            )));
+        }
 
-        let block = blocks.into_iter().next().expect("block to exist");
-        let header = block.header().clone();
-
-        let result = match self
+        match self
             .chainstate_handle
-            .call(move |this| this.preliminary_block_check(block))
+            .call_mut(|c| {
```

Comment:
It's an open question whether we want to first run `call` and do a check on the block before calling the with `call_mut` and `process_block`. The reasoning is that `call` can run in parallel, hence an attacker can't block our subsystems.